Copyright (c) Microsoft Corporation. All rights reserved.  
Licensed under the MIT License.

![Impressions](https://PixelServer20190423114238.azurewebsites.net/api/impressions/NotebookVM/how-to-use-azureml/machine-learning-pipelines/intro-to-pipelines/aml-pipelines-with-automated-machine-learning-step.png)

# Azure Machine Learning Pipeline with AutoMLStep (Udacity Course 2)
This notebook demonstrates the use of AutoMLStep in Azure Machine Learning Pipeline.

## Introduction
In this example we showcase how you can use AzureML Dataset to load data for AutoML via AML Pipeline. 

If you are using an Azure Machine Learning Notebook VM, you are all set. Otherwise, make sure you have executed the [configuration](https://aka.ms/pl-config) before running this notebook.

In this notebook you will learn how to:
1. Create an `Experiment` in an existing `Workspace`.
2. Create or Attach existing AmlCompute to a workspace.
3. Define data loading in a `TabularDataset`.
4. Configure AutoML using `AutoMLConfig`.
5. Use AutoMLStep
6. Train the model using AmlCompute
7. Explore the results.
8. Test the best fitted model.

## Azure Machine Learning and Pipeline SDK-specific imports

In [1]:
import logging
import os
import csv

from matplotlib import pyplot as plt
import numpy as np
import pandas as pd
from sklearn import datasets
import pkg_resources

import azureml.core
from azureml.core.experiment import Experiment
from azureml.core.workspace import Workspace
from azureml.train.automl import AutoMLConfig
from azureml.core.dataset import Dataset

from azureml.pipeline.steps import AutoMLStep

# Check core SDK version number
print("SDK version:", azureml.core.VERSION)

SDK version: 1.23.0


## Initialize Workspace
Initialize a workspace object from persisted configuration. Make sure the config file is present at .\config.json

In [2]:
ws = Workspace.from_config()
print(ws.name, ws.resource_group, ws.location, ws.subscription_id, sep = '\n')

Note, we have launched a browser for you to login. For old experience with device code, use "az login --use-device-code"
Performing interactive authentication. Please follow the instructions on the terminal.
You have logged in. Now let us find all the subscriptions to which you have access...
Interactive authentication successfully completed.
quick-starts-ws-139987
aml-quickstarts-139987
southcentralus
cdbe0b43-92a0-4715-838a-f2648cc7ad21


## Create an Azure ML experiment
Let's create an experiment named "automlstep-classification" and a folder to hold the training scripts. The script runs will be recorded under the experiment in Azure.

The best practice is to use separate folders for scripts and its dependent files for each step and specify that folder as the `source_directory` for the step. This helps reduce the size of the snapshot created for the step (only the specific folder is snapshotted). Since changes in any files in the `source_directory` would trigger a re-upload of the snapshot, this helps keep the reuse of the step when there are no changes in the `source_directory` of the step.

*Udacity Note:* There is no need to create an Azure ML experiment, this needs to re-use the experiment that was already created


In [3]:
# Choose a name for the run history container in the workspace.
# NOTE: update these to match your existing exp name
experiment_name = 'automl-bank-experiment'
project_folder = './pipeline-bank-project'

experiment = Experiment(ws, experiment_name)

experiment

Experiment(Name: automl-bank-experiment,
Workspace: quick-starts-ws-139987)

In [4]:

dic_data = {'Workspace name': ws.name,
            'Azure region': ws.location,
            'Subscription id': ws.subscription_id,
            'Resource group': ws.resource_group,
            'Experiment Name': experiment.name}

df_data = pd.DataFrame.from_dict(data = dic_data, orient='index')

df_data.rename(columns={0:''}, inplace = True)
df_data

Workspace name                 quick-starts-ws-139987
Azure region                           southcentralus
Subscription id  cdbe0b43-92a0-4715-838a-f2648cc7ad21
Resource group                 aml-quickstarts-139987
Experiment Name                automl-bank-experiment

### Create or Attach an AmlCompute cluster
You will need to create a [compute target](https://docs.microsoft.com/azure/machine-learning/service/concept-azure-machine-learning-architecture#compute-target) for your AutoML run. In this tutorial, you get the default `AmlCompute` as your training compute resource.

**Udacity Note** There is no need to create a new compute target, it can re-use the previous cluster

In [5]:
from azureml.core.compute import AmlCompute
from azureml.core.compute import ComputeTarget
from azureml.core.compute_target import ComputeTargetException

# NOTE: update the cluster name to match the existing cluster
# Choose a name for your CPU cluster
amlcompute_cluster_name = "cpu-cluster"

# Verify that cluster does not exist already
try:
    compute_target = ComputeTarget(workspace=ws, name=amlcompute_cluster_name)
    print('Found existing cluster, use it.')
except ComputeTargetException:
    compute_config = AmlCompute.provisioning_configuration(vm_size='STANDARD_DS12_V2',# for GPU, use "STANDARD_NC6"
                                                           #vm_priority = 'lowpriority', # optional
                                                           min_nodes=1,
                                                           max_nodes=4)
    compute_target = ComputeTarget.create(ws, amlcompute_cluster_name, compute_config)
    
    compute_target.wait_for_completion(show_output=True, min_node_count = 1, timeout_in_minutes = 10)
# For a more detailed view of current AmlCompute status, use get_status().
print(compute_target.get_status().serialize())

Creating...
SucceededProvisioning operation finished, operation "Succeeded"
Succeeded..................
AmlCompute wait for completion finished

Minimum number of nodes requested have been provisioned
{'currentNodeCount': 1, 'targetNodeCount': 1, 'nodeStateCounts': {'preparingNodeCount': 1, 'runningNodeCount': 0, 'idleNodeCount': 0, 'unusableNodeCount': 0, 'leavingNodeCount': 0, 'preemptedNodeCount': 0}, 'allocationState': 'Steady', 'allocationStateTransitionTime': '2021-03-07T10:56:21.788000+00:00', 'errors': None, 'creationTime': '2021-03-07T10:54:28.568847+00:00', 'modifiedTime': '2021-03-07T10:54:43.917406+00:00', 'provisioningState': 'Succeeded', 'provisioningStateTransitionTime': None, 'scaleSettings': {'minNodeCount': 1, 'maxNodeCount': 4, 'nodeIdleTimeBeforeScaleDown': 'PT120S'}, 'vmPriority': 'Dedicated', 'vmSize': 'STANDARD_DS12_V2'}


In [6]:
# Check details about compute_targets (i.e. compute_target)

compute_targets = ws.compute_targets
for name, ct in compute_targets.items():
    print(name, ct.type, ct.provisioning_state)

cpu-cluster AmlCompute Succeeded


## Data

**Udacity note:** Make sure the `key` is the same name as the dataset that is uploaded, and that the description matches. If it is hard to find or unknown, loop over the `ws.datasets.keys()` and `print()` them.
If it *isn't* found because it was deleted, it can be recreated with the link that has the CSV 

In [7]:
# Try to load the dataset from the Workspace. Otherwise, create it from the file
# NOTE: update the key to match the dataset name
found = False
key = "BankMarketing Dataset"
description_text = "Bank Marketing DataSet for Udacity Course 2"

if key in ws.datasets.keys(): 
        found = True
        dataset = ws.datasets[key] 

if not found:
        # Create AML Dataset and register it into Workspace
        example_data = 'https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv'
        dataset = Dataset.Tabular.from_delimited_files(example_data)        
        #Register Dataset in Workspace
        dataset = dataset.register(workspace=ws,
                                   name=key,
                                   description=description_text)


df = dataset.to_pandas_dataframe()

# Describe dataset
m, k = df.shape
print("{} x {} table of data:".format(m, k))
display(df.head())
print("...")

32950 x 21 table of data:


age          job  marital    education  default housing loan    contact  \
0   57   technician  married  high.school       no      no  yes   cellular   
1   55      unknown  married      unknown  unknown     yes   no  telephone   
2   33  blue-collar  married     basic.9y       no      no   no   cellular   
3   36       admin.  married  high.school       no      no   no  telephone   
4   27    housemaid  married  high.school       no     yes   no   cellular   

  month day_of_week  ...  campaign  pdays  previous     poutcome emp.var.rate  \
0   may         mon  ...         1    999         1      failure         -1.8   
1   may         thu  ...         2    999         0  nonexistent          1.1   
2   may         fri  ...         1    999         1      failure         -1.8   
3   jun         fri  ...         4    999         0  nonexistent          1.4   
4   jul         fri  ...         2    999         0  nonexistent          1.4   

   cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y  
0          92.893          -46.2      1.299       5099.1  no  
1          93.994          -36.4      4.860       5191.0  no  
2          92.893          -46.2      1.313       5099.1  no  
3          94.465          -41.8      4.967       5228.1  no  
4          93.918          -42.7      4.963       5228.1  no  

[5 rows x 21 columns]

...


### Further review of the Dataset

In [8]:
df.y.value_counts(normalize=True)

no     0.887951
yes    0.112049
Name: y, dtype: float64

We are facing a case of `Imbalanced data`: **88.80% - no** and **11.20% - yes**.

**Automated ML** has built in capabilities to help deal with imbalanced data such as:

* Apply a weight column to the dataset.
* Run an experiment with sub-sampled data.
* Review performance metric for imbalanced data such as AUC_weighted and F1-score.
* Resampling to even the class imbalance.

## Train
This creates a general AutoML settings object.
**Udacity notes:** These inputs must match what was used when training in the portal. `label_column_name` has to be `y` for example.

In [9]:
automl_settings = {
    "experiment_timeout_minutes": 60, # define the duration of the experiment (in minutes).
    "max_concurrent_iterations": 5,
    "primary_metric" : 'AUC_weighted'
}
automl_config = AutoMLConfig(compute_target=compute_target,
                             task = "classification",
                             training_data=dataset,
                             label_column_name="y",   
                             path = project_folder,
                             enable_early_stopping= True,
                             featurization= 'auto',
                             debug_log = "automl_errors.log",
                             #blocked_models=['XGBoostClassifier'],
                             **automl_settings
                            )

**__NOTE__:** I am using the parameter `blocked_models=['XGBoostClassifier']` in my `AutoMLConfig` because:

* AutoML uses `Python 3.6` & `py-xgboost==0.90`.
* My Python dev environment uses `Python 3.8.5` which does not support `py-xgboost<=0.90`.

#### Create Pipeline and AutoMLStep

You can define outputs for the AutoMLStep using TrainingOutput.

In [10]:
from azureml.pipeline.core import PipelineData, TrainingOutput

ds = ws.get_default_datastore()
metrics_output_name = 'metrics_output'
best_model_output_name = 'best_model_output'

metrics_data = PipelineData(name='metrics_data',
                           datastore=ds,
                           pipeline_output_name=metrics_output_name,
                           training_output=TrainingOutput(type='Metrics'))
model_data = PipelineData(name='model_data',
                           datastore=ds,
                           pipeline_output_name=best_model_output_name,
                           training_output=TrainingOutput(type='Model'))

Create an `AutoMLStep`.

In [11]:
automl_step = AutoMLStep(
    name='automl_module',
    automl_config=automl_config,
    outputs=[metrics_data, model_data],
    allow_reuse=True)

We can now create a pipeline using the `AutoMLStep` created above.

In [12]:
from azureml.pipeline.core import Pipeline
pipeline = Pipeline(
    description="pipeline_with_automlstep",
    workspace=ws,    
    steps=[automl_step])

In [13]:
pipeline_run = experiment.submit(pipeline)

Created step automl_module [b4172de6][4ccd3676-bc94-49ac-9858-c992efc421a9], (This step will run and generate new outputs)
Submitted PipelineRun 710d17b1-ae26-4674-868f-ed19faf8741d
Link to Azure Machine Learning Portal: https://ml.azure.com/experiments/automl-bank-experiment/runs/710d17b1-ae26-4674-868f-ed19faf8741d?wsid=/subscriptions/cdbe0b43-92a0-4715-838a-f2648cc7ad21/resourcegroups/aml-quickstarts-139987/workspaces/quick-starts-ws-139987


In [14]:
from azureml.widgets import RunDetails
RunDetails(pipeline_run).show()

_PipelineWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', …

In [15]:
pipeline_run.wait_for_completion()

PipelineRunId: 710d17b1-ae26-4674-868f-ed19faf8741d
Link to Azure Machine Learning Portal: https://ml.azure.com/experiments/automl-bank-experiment/runs/710d17b1-ae26-4674-868f-ed19faf8741d?wsid=/subscriptions/cdbe0b43-92a0-4715-838a-f2648cc7ad21/resourcegroups/aml-quickstarts-139987/workspaces/quick-starts-ws-139987
PipelineRun Status: Running


StepRunId: c06b7d5c-491f-4021-a6e0-9408dfa83d74
Link to Azure Machine Learning Portal: https://ml.azure.com/experiments/automl-bank-experiment/runs/c06b7d5c-491f-4021-a6e0-9408dfa83d74?wsid=/subscriptions/cdbe0b43-92a0-4715-838a-f2648cc7ad21/resourcegroups/aml-quickstarts-139987/workspaces/quick-starts-ws-139987
StepRun( automl_module ) Status: NotStarted
StepRun( automl_module ) Status: Running

StepRun(automl_module) Execution Summary
StepRun( automl_module ) Status: Finished
{'runId': 'c06b7d5c-491f-4021-a6e0-9408dfa83d74', 'target': 'cpu-cluster', 'status': 'Completed', 'startTimeUtc': '2021-03-07T11:08:09.510303Z', 'endTimeUtc': '2021-03-0

'Finished'

## Examine Results

### Retrieve the metrics of all child runs
Outputs of above run can be used as inputs of other steps in pipeline. In this tutorial, we will examine the outputs by retrieve output data and running some tests.

In [16]:
metrics_output = pipeline_run.get_pipeline_output(metrics_output_name)
num_file_downloaded = metrics_output.download('.', show_progress=True)

Downloaded azureml/c06b7d5c-491f-4021-a6e0-9408dfa83d74/metrics_data, 1 files out of an estimated total of 1


In [17]:
import json
with open(metrics_output._path_on_datastore) as f:
    metrics_output_result = f.read()
    
deserialized_metrics_output = json.loads(metrics_output_result)
df = pd.DataFrame(deserialized_metrics_output)
df

[0.5067488093455327]   
accuracy                                            [0.9101669195751139]   
f1_score_micro                                      [0.9101669195751139]   
recall_score_micro                                  [0.9101669195751139]   
recall_score_weighted                               [0.9101669195751139]   
average_precision_score_micro                       [0.9791160013753892]   
AUC_micro                                           [0.9780521828033002]   
AUC_weighted                                        [0.9400218950924984]   
log_loss                                           [0.18503224281252317]   
precision_score_micro                               [0.9101669195751139]   
average_precision_score_weighted                    [0.9500290713482653]   
average_precision_score_macro                        [0.803381929914819]   
weighted_accuracy                                    [0.955397425760583]   
f1_score_weighted                                    [0.905449408966568]   
precision_score_macro                               [0.7815896101875621]   
norm_macro_recall                                  [0.45597271078657475]   
recall_score_macro                                  [0.7279863553932874]   
precision_score_weighted                             [0.902773605802737]   
f1_score_macro                                      [0.7507997342735959]   
AUC_macro                                           [0.9400218950924984]   

                                 c06b7d5c-491f-4021-a6e0-9408dfa83d74_38  ...  \
balanced_accuracy                                   [0.7396688320950195]  ...   
matthews_correlation                                 [0.523106107160623]  ...   
accuracy                                            [0.9119878603945372]  ...   
f1_score_micro                                      [0.9119878603945372]  ...   
recall_score_micro                                  [0.9119878603945372]  ...   
recall_score_weighted                               [0.9119878603945372]  ...   
average_precision_score_micro                       [0.9794493482917321]  ...   
AUC_micro                                            [0.978414897266977]  ...   
AUC_weighted                                        [0.9407813695361834]  ...   
log_loss                                           [0.18438902755770786]  ...   
precision_score_micro                               [0.9119878603945372]  ...   
average_precision_score_weighted                    [0.9509461669750866]  ...   
average_precision_score_macro                       [0.8071742282912644]  ...   
weighted_accuracy                                   [0.9547700139704612]  ...   
f1_score_weighted                                   [0.9081566096944386]  ...   
precision_score_macro                               [0.7854355288470023]  ...   
norm_macro_recall                                   [0.4793376641900391]  ...   
recall_score_macro                                  [0.7396688320950195]  ...   
precision_score_weighted                            [0.9057755480984142]  ...   
f1_score_macro                                      [0.7597246696921418]  ...   
AUC_macro                                           [0.9407813695361835]  ...   

                                 c06b7d5c-491f-4021-a6e0-9408dfa83d74_26  \
balanced_accuracy                                   [0.7459312545962097]   
matthews_correlation                                 [0.530535514228425]   
accuracy                                            [0.9125948406676783]   
f1_score_micro                                      [0.9125948406676783]   
recall_score_micro                                  [0.9125948406676783]   
recall_score_weighted                               [0.9125948406676783]   
average_precision_score_micro                       [0.9800058297942514]   
AUC_micro                                           [0.9790132195513965]   
AUC_weighted                                        [0.942311432

### Retrieve the Best Model

If you are a Mac user make sure you have:

* **lightgbm** installed: `brew install lightgbm`
* Confirm the version of **XGBoost** installed on your local environment matches with the installed on the AutoML training environment.

Install the correct version as follows:
```bash
conda install -c anaconda py-xgboost==<the_specific_version>
```


In [18]:
# Retrieve best model from Pipeline Run
best_model_output = pipeline_run.get_pipeline_output(best_model_output_name)
num_file_downloaded = best_model_output.download('.', show_progress=True)

Downloaded azureml/c06b7d5c-491f-4021-a6e0-9408dfa83d74/model_data, 1 files out of an estimated total of 1


In [19]:
from azureml.core.environment import Environment

# Retrieve AutoML training environment for the best model
print(Environment.get(ws, "AzureML-AutoML").python.conda_dependencies.serialize_to_string())

channels:
- anaconda
- conda-forge
- pytorch
dependencies:
- python=3.6.2
- pip=20.2.4
- pip:
  - azureml-core==1.23.0
  - azureml-pipeline-core==1.23.0
  - azureml-telemetry==1.23.0
  - azureml-defaults==1.23.0
  - azureml-interpret==1.23.0
  - azureml-automl-core==1.23.0
  - azureml-automl-runtime==1.23.0
  - azureml-train-automl-client==1.23.0
  - azureml-train-automl-runtime==1.23.0
  - azureml-dataset-runtime==1.23.0
  - azureml-mlflow==1.23.0
  - inference-schema
  - py-cpuinfo==5.0.0
  - boto3==1.15.18
  - botocore==1.18.18
- numpy~=1.18.0
- scikit-learn==0.22.1
- pandas~=0.25.0
- py-xgboost<=0.90
- fbprophet==0.5
- holidays==0.9.11
- setuptools-git
- psutil>5.0.0,<6.0.0
name: azureml_661474bbe74e96b5d8added5888dfc85



In [20]:
import pickle

with open(best_model_output._path_on_datastore, "rb" ) as f:
    best_model = pickle.load(f)
best_model

PipelineWithYTransformations(Pipeline={'memory': None,
                                       'steps': [('datatransformer',
                                                  DataTransformer(enable_dnn=None,
                                                                  enable_feature_sweeping=None,
                                                                  feature_sweeping_config=None,
                                                                  feature_sweeping_timeout=None,
                                                                  featurization_config=None,
                                                                  force_text_dnn=None,
                                                                  is_cross_validation=None,
                                                                  is_onnx_compatible=None,
                                                                  logger=None,
                                                              

In [21]:
best_model.steps

[('datatransformer',
  DataTransformer(enable_dnn=None, enable_feature_sweeping=None,
                  feature_sweeping_config=None, feature_sweeping_timeout=None,
                  featurization_config=None, force_text_dnn=None,
                  is_cross_validation=None, is_onnx_compatible=None, logger=None,
                  observer=None, task=None, working_dir=None)),
 ('prefittedsoftvotingclassifier',
  PreFittedSoftVotingClassifier(classification_labels=None,
                                estimators=[('31',
                                             Pipeline(memory=None,
                                                      steps=[('standardscalerwrapper',
                                                              <azureml.automl.runtime.shared.model_wrappers.StandardScalerWrapper object at 0x7fa310290358>),
                                                             ('xgboostclassifier',
                                                              XGBoostClassifier(ba

### Test the Model
#### Load Test Data
For the test data, it should have the same preparation step as the train data. Otherwise it might get failed at the preprocessing step.

In [22]:
dataset_test = Dataset.Tabular.from_delimited_files(path='https://automlsamplenotebookdata.blob.core.windows.net/automl-sample-notebook-data/bankmarketing_train.csv')
df_test = dataset_test.to_pandas_dataframe()
df_test = df_test[pd.notnull(df_test['y'])]

y_test = df_test['y']
X_test = df_test.drop(['y'], axis=1)

# Describe dataset
m, k = X_test.shape
print("{} x {} table of data:".format(m, k))
display(X_test.head())
print("...")

32950 x 20 table of data:


age          job  marital    education  default housing loan    contact  \
0   57   technician  married  high.school       no      no  yes   cellular   
1   55      unknown  married      unknown  unknown     yes   no  telephone   
2   33  blue-collar  married     basic.9y       no      no   no   cellular   
3   36       admin.  married  high.school       no      no   no  telephone   
4   27    housemaid  married  high.school       no     yes   no   cellular   

  month day_of_week  duration  campaign  pdays  previous     poutcome  \
0   may         mon       371         1    999         1      failure   
1   may         thu       285         2    999         0  nonexistent   
2   may         fri        52         1    999         1      failure   
3   jun         fri       355         4    999         0  nonexistent   
4   jul         fri       189         2    999         0  nonexistent   

   emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed  
0         -1.80           92.89         -46.20       1.30      5099.10  
1          1.10           93.99         -36.40       4.86      5191.00  
2         -1.80           92.89         -46.20       1.31      5099.10  
3          1.40           94.47         -41.80       4.97      5228.10  
4          1.40           93.92         -42.70       4.96      5228.10

...


#### Testing Our Best Fitted Model

We will use confusion matrix to see how our model works.

In [23]:
from sklearn.metrics import confusion_matrix
ypred = best_model.predict(X_test)
cm = confusion_matrix(y_test, ypred)

In [24]:
# Visualize the confusion matrix
pd.DataFrame(cm).style.background_gradient(cmap='Blues', low=0, high=0.9)

## Publish and run from REST endpoint

Run the following code to publish the pipeline to your workspace. In your workspace in the portal, you can see metadata for the pipeline including run history and durations. You can also run the pipeline manually from the portal.

Additionally, publishing the pipeline enables a REST endpoint to rerun the pipeline from any HTTP library on any platform.


In [25]:
published_pipeline = pipeline_run.publish_pipeline(
    name="Bankmarketing Train", 
    description="Training bankmarketing pipeline", 
    version="1.0")

published_pipeline


Pipeline(Name: Bankmarketing Train,
Id: 13c5a0c1-6da0-4bf5-a0cf-8cd2467465cb,
Status: Active,
Endpoint: https://southcentralus.api.azureml.ms/pipelines/v1.0/subscriptions/cdbe0b43-92a0-4715-838a-f2648cc7ad21/resourceGroups/aml-quickstarts-139987/providers/Microsoft.MachineLearningServices/workspaces/quick-starts-ws-139987/PipelineRuns/PipelineSubmit/13c5a0c1-6da0-4bf5-a0cf-8cd2467465cb)

Authenticate once again, to retrieve the `auth_header` so that the endpoint can be used

In [26]:
from azureml.core.authentication import InteractiveLoginAuthentication

interactive_auth = InteractiveLoginAuthentication()
auth_header = interactive_auth.get_authentication_header()



Get the REST url from the endpoint property of the published pipeline object. You can also find the REST url in your workspace in the portal. Build an HTTP POST request to the endpoint, specifying your authentication header. Additionally, add a JSON payload object with the experiment name and the batch size parameter. As a reminder, the process_count_per_node is passed through to ParallelRunStep because you defined it is defined as a PipelineParameter object in the step configuration.

Make the request to trigger the run. Access the Id key from the response dict to get the value of the run id.


In [27]:
import requests

rest_endpoint = published_pipeline.endpoint
response = requests.post(rest_endpoint, 
                         headers=auth_header, 
                         json={"ExperimentName": "pipeline-rest-endpoint"}
                        )

In [28]:
try:
    response.raise_for_status()
except Exception:    
    raise Exception("Received bad response from the endpoint: {}\n"
                    "Response Code: {}\n"
                    "Headers: {}\n"
                    "Content: {}".format(rest_endpoint, response.status_code, response.headers, response.content))

run_id = response.json().get('Id')
print('Submitted pipeline run: ', run_id)

Submitted pipeline run:  38d41713-c782-4721-851e-43d85030db42


Use the run id to monitor the status of the new run. This will take another 10-15 min to run and will look similar to the previous pipeline run, so if you don't need to see another pipeline run, you can skip watching the full output.

In [29]:
from azureml.pipeline.core.run import PipelineRun
from azureml.widgets import RunDetails

published_pipeline_run = PipelineRun(ws.experiments["pipeline-rest-endpoint"], run_id)
RunDetails(published_pipeline_run).show()

_PipelineWidget(widget_settings={'childWidgetDisplay': 'popup', 'send_telemetry': False, 'log_level': 'INFO', …